File: py_tf2_tutorial/py_tf2_tutorial/tf2_add_frame.py

import rclpy
import numpy as np
from rclpy.node import Node

from tf2_ros import TransformBroadcaster
from geometry_msgs.msg import TransformStamped

class InverOrbitFrame(Node):

   def __init__(self):
        super().__init__('inverse_orbit_frame_tf2_broadcaster')
        
        self.count = 0
        self.freq_factor = 2* np.pi / 20;

        self.tf_broadcaster = TransformBroadcaster(self)
        self.timer = self.create_timer(0.1, self.broadcast_timer_callback)

   def broadcast_timer_callback(self):
       
       t = TransformStamped()

       t.header.stamp = self.get_clock().now().to_msg()
       t.header.frame_id = 'animated_sphere'
       t.child_frame_id = 'inverse_orbit_frame'

       t.transform.translation.x = np.cos(-self.count * self.freq_factor)
       t.transform.translation.y = np.sin(-self.count * self.freq_factor)
       t.transform.translation.z = 0.5

       t.transform.rotation.x = 0.0
       t.transform.rotation.y = 0.0
       t.transform.rotation.z = 0.0
       t.transform.rotation.w = 1.0

       self.tf_broadcaster.sendTransform(t)
       self.count += 1
       if self.count >= 20:
           self.count = 0

def main():
    rclpy.init()
    node = InverOrbitFrame()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass

    rclpy.shutdown()

File: cpp_tf2_tutorial/src/tf2_add_frame.cpp
#include "rclcpp/rclcpp.hpp"
#include "tf2_ros/transform_broadcaster.h"
#include "geometry_msgs/msg/transform_stamped.hpp"

const double PI = 3.14159265358979323846;

using namespace std::chrono_literals;

class InverOrbitFrame : public rclcpp::Node
{
private:
  int count = 0;
  double freq_factor = 2*PI/20;
  rclcpp::TimerBase::SharedPtr timer_;
  std::shared_ptr<tf2_ros::TransformBroadcaster> m_tf_broadcaster;
public:
  InverOrbitFrame() : Node("inverse_orbit_frame_tf2_broadcaster"){
    m_tf_broadcaster = std::make_shared<tf2_ros::TransformBroadcaster>(this);
    timer_ = this->create_wall_timer(
      100ms, std::bind(&InverOrbitFrame::broadcast_timer_callback, this));
  }

  void broadcast_timer_callback(){
    geometry_msgs::msg::TransformStamped t;

    t.header.stamp = this->get_clock()->now();
    t.header.frame_id = "animated_sphere";
    t.child_frame_id = "inverse_orbit_frame";
    
    t.transform.translation.x = cos(-freq_factor * count);
    t.transform.translation.y = sin(-freq_factor * count);
    t.transform.translation.z = 0.5;

    t.transform.rotation.x = 0.0;
    t.transform.rotation.y = 0.0;
    t.transform.rotation.z = 0.0;
    t.transform.rotation.w = 1.0;

    m_tf_broadcaster->sendTransform(t);
    count++;
    count >= 20 ? count = 0 : count;
  }
};

int main(int argc, char * argv[]){
  rclcpp::init(argc, argv);

  rclcpp::spin(std::make_shared<InverOrbitFrame>());

  rclcpp::shutdown();
  return 0;
}